onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc

```diff
@@ -2304,6 +2346,88 @@ void DequantizeLinearOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builde
 
 #pragma endregion
 
+#pragma region op_LRN
+
+class LRNOpBuilder : public BaseOpBuilder {
+ private:
+  bool IsOpSupportedImpl(ModelBuilder& model_builder, const Node& node) override;
+
+  int32_t GetMinSupportedSdkVer(ModelBuilder& /* model_builder */, const Node& /* node */) const override {
+    return 28;
+  }
+
+  void AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node) override;
+};
+
+bool LRNOpBuilder::IsOpSupportedImpl(ModelBuilder& /* model_builder */, const Node& node) {
+  Shape input_shape;
+  if (!GetShape(*node.InputDefs()[0], input_shape))
+    return false;
+
+  const auto input_size = input_shape.size();
+  if (input_size != 4) {
+    LOGS_DEFAULT(VERBOSE) << "LRN only support 4d shape, input is "
+                          << input_size << "d shape";
+    return false;
+  }
+
```

Comment:
If the model has no batch dimension, can we support 3D and add 1 for the batch size as the first dimension?

Reply:
Actually doesn't matter. The CPU implementation hardcodes a requirement of 4 and the CUDA implementation requires 4 or 5.

---
In reply to: [459175160](https://github.com/microsoft/onnxruntime/pull/4582#discussion_r459175160) [](ancestors = 459175160)

Reply:
ONNX LRN supports 4+ dimension input, while NNAPI ANEURALNETWORKS_LOCAL_RESPONSE_NORMALIZATION supports only 4D input